compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,175 @@
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/GlobalOptimization/PassDetail.h"
+#include "iree/compiler/GlobalOptimization/Passes.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/PatternMatch.h"
+
+#include "llvm/Support/Debug.h"
+
+using namespace mlir;
+
+#define DEBUG_TYPE "global-loop-invariant-code-motion"
+#define LICM_DBGS() (llvm::dbgs() << '[' << DEBUG_TYPE << "] ")
+
+// Check if the op is a leaf op we always want to hoist.
+static bool isHoistableLeafOp(Operation *op) {
+  // Currently it's limited to a small set of ops related to constant pack op.
+  return isa<tensor::PackOp>(op);
+}
+
+// Check if the op is hoistable (but we might not want to hoist it alone).
+static bool isHoistableOp(Operation *op) {
+  // Currently it's limited to a small set of ops related to constant pack op.
+  return op->hasTrait<OpTrait::ConstantLike>() || isa<tensor::EmptyOp>(op) ||
+         isHoistableLeafOp(op);
+}
+
+// Check if the op and its producers are loop invariants and hoistable. Results
+// are cached in hoistableOpMap to avoid repeated traversals.
+static bool checkHoistableBackwardSlice(
+    LoopLikeOpInterface loopOp, Operation *op,
+    llvm::SmallDenseMap<Operation *, bool> &hoistableOpMap) {
+  // First check if the op has been analyzed.
+  if (hoistableOpMap.contains(op))
+    return hoistableOpMap[op];
+
+  bool hoistable = true;
+  // Currently only hoist ops with no region (so no implicit capture).
+  if (op->getNumRegions() > 0) {
+    hoistable = false;
+  } else {
+    // Check if all producers are hoistable.
+    for (OpOperand &operand : op->getOpOperands()) {
+      Value value = operand.get();
+      // Ignore values defined outside the loop.
+      if (loopOp.isDefinedOutsideOfLoop(value))
+        continue;
+
+      Operation *producer = value.getDefiningOp();
+      // If the value is not an operation, we don't hoist it.
+      if (!producer ||
+          !checkHoistableBackwardSlice(loopOp, producer, hoistableOpMap)) {
+        hoistable = false;
+        break;
+      }
+    }
+  }
+
+  hoistableOpMap[op] = hoistable;
+
+  LLVM_DEBUG(LICM_DBGS() << (hoistable ? "Hoistable: " : "Non-hoistable: ")
+                         << *op << "\n");
+  return hoistable;
+}
+
+// Call `moveOutOfLoop` to hoist op and its producer out of the loop. Ops are
+// hoisted in post-order to handle the dependencies (leaves of the producer tree
+// are hoisted first).
+static LogicalResult hoistBackwardSlice(Operation *op,
+                                        LoopLikeOpInterface loopOp) {
+  // Check if the op has been hoisted.
+  if (!loopOp->isAncestor(op))
+    return success();
+
+  // Walk all producers and hoist them first.
```

Comment:
This kind of lamda recursing usage to recurse is very hard to follow. These all seem like they need to be worklists

Reply:
IIUC the tricky part here is to hoist them in topological order. The recursive is following that naturally. To use a worklist we need to collect all ops and do a topological sorting first, which seems to be more complicated to me.

This code is also inspired by how IREE currently recursively hoists expression tree into globals: 

https://github.com/openxla/iree/blob/0c61f779bf9e56938a66625d214ab36fb85f684c/compiler/src/iree/compiler/Dialect/Util/Transforms/HoistIntoGlobals.cpp#L246-L257

Reply:
Had an offline discussion with @dcaballe . I changed the hoisting function to use worklist to gather backward slices and sort all ops to be hoisted with `mlir::computeTopologicalSorting`, then hoist them in topological order.

This should resolve dependency issue and avoid recursion, while not making code too complicated.